### PR TITLE
Add Vue support

### LINKED ISSUE #162: Vue support
Unfortunately this doesn't seem to detect Typescript code in `.vue` files out of the box.

To reproduce, create a basic Vue project (Vue CLI can do this), and then export something from a `.ts` file and import it in a `.vue` file. `ts-prune` will report it as unused.

My `tsconfig.json` file does include `.vue` files:

```
  "include": [
    ...
    "src/**/*.vue",
  ],
```

However I guess it doesn't understand Vue single file components. They are converted to Typescript during webpack compilation by the `vue-loader` plugin.

### PR DESCRIPTION
**Pull request by Timmmm**
This also adds a test that may or may not test anything.

Fixes #162

### CODE REVIEW

File: example/vue/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "compilerOptions": {
+    "target": "ES5",
+    "module": "commonjs",
+    "noEmit": true,
+    "noImplicitAny": true,
+    "strictNullChecks": true
+  },
+  "include": ["./**/*.ts", "./**/*.vue"],
+  "atom": {
```
Comment (mrseanryan) [inline example/vue/tsconfig.json:10]: Remove `atom` section

File: src/app.ts [hunk 2]
```diff
@@ -23,6 +23,16 @@ const parseTsConfig = (tsconfigPath: string): TsConfig => {
...
       ts.sys,
       basePath,
+      undefined,
+      undefined,
+      undefined,
+      [
+        {
+          extension: 'vue',
+          isMixedContent: true,
+          scriptKind: ts.ScriptKind.Deferred,
```
Comment (mrseanryan) [inline src/app.ts:33]: I wonder could this cause any issues.

For example if an existing user has some `vue` files that they do NOT want scanned.

Maybe this part should be behind a new option like `--vue` - then that could also auto-suppress the shims file `shims-tsx-d-ts`

**Review by mrseanryan — CHANGES_REQUESTED** [review]
Nice work - a few TODOs:

________________________________________
A) README

TODO - add note to README on Vue support. Something like this:

_____________________

VUE Support

1. To process VUE files, your tsconfig.json should include the vue files:

```
"include": ["./**/*.ts", "./**/*.vue"],
```

2. Suppress warnings about the `shims-vue.d.ts` file (default export from Vue components).   To do this, use the `ignoreFiles` option:

```
--ignoreFiles=shims-vue.d.ts
```

________________________________________
B) CHANGELOG

Please add an 'Added' entry to the top of the CHANGELOG.md file

File: src/app.ts (hunk 2 above)
Comment (mrseanryan) [reply src/app.ts:33]: Then, the README change is much smaller  (just document the new option)

Comment (mrseanryan) [reply src/app.ts:33]: I prefer the approach of adding a new option `--vue` 

@pzavolinsky what do you reckon ?

**Comment by mrseanryan** [comment]
(related) TODO @mrseanryan check what happens if more than 3 issues found -> itest should fail ...

File: example/vue/math.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export function add1(x: number) { return x + 1; }
```
Comment (mrseanryan) [inline example/vue/math.ts:1]: Rename add1 to unused

File: example/vue/math.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+
+// ts-unused-exports:disable-next-line
+export function add2(x: number) { return x + 2; }
```
Comment (mrseanryan) [inline example/vue/math.ts:4]: Rename add2 to unusedButDisabled

**Review by mrseanryan — CHANGES_REQUESTED** [review]

**Comment by mrseanryan** [comment]
TODO add itest with more complex embedded Typescript.

TODO @Timmmm what about Mixins? Looks like something that could require special support.

Ref = https://blog.logrocket.com/how-to-write-a-vue-js-app-completely-in-typescript/

Are there any other Vue features we need to think about?

File: src/app.ts (hunk 2 above)
Comment (mrseanryan) [inline src/app.ts:33]: Please add comme t to explain the values used for:

isMixedContent,
          scriptKind

**Review by mrseanryan — CHANGES_REQUESTED** [review]
Comments needed to explain parameters chosen.

Concern about advanced Vue features like Mixins

**Comment by mrseanryan** [comment]
TODO add a new **unit test** file

The PR has an itest, but some unit tests are needed for future maintenance and debugging. 

Add a new file like: vue.feature

Example:

https://github.com/pzavolinsky/ts-unused-exports/blob/master/features/base-url-undefined.feature

File: example/vue/tsconfig.json (hunk 1 above)
Comment (Timmmm) [reply example/vue/tsconfig.json:10]: Hmm yeah not sure why this is in here - I just copied it from another example, you might want to delete it there too.

File: example/vue/math.ts (hunk 3 above)
Comment (Timmmm) [reply example/vue/math.ts:1]: These functions were just copied from another example, I left the names for consistency.

**Comment by Timmmm** [comment]
> what about Mixins? Looks like something that could require special support.

I think mixins are [a standard Typescript feature](https://www.typescriptlang.org/docs/handbook/mixins.html).

**Comment by mrseanryan** [comment]
@Timmmm about the itest not failing:

The itests use the `exitWithCount` option:

> Set the process exit code to be the count of files that have unused exports.     

So it is not counting the number of unused items - it is counting the *files*.

When I tried it on this branch, it worked ok (by making a copy of Math.ts)

File: example/vue/tsconfig.json (hunk 1 above)
Comment (mrseanryan) [reply example/vue/tsconfig.json:10]: Probably, but we can start by removing it here ...

Comment (mrseanryan) [reply example/vue/tsconfig.json:10]: Logged #172 but if we can some of it here that would be no harm, as adding new files.

Comment (mrseanryan) [reply example/vue/tsconfig.json:10]: #172 has been fixed

File: example/vue/math.ts (hunk 3 above)
Comment (mrseanryan) [reply example/vue/math.ts:1]: Fixed in master - if you rebase you will see.

**Comment by FlorianWendelborn** [comment]
Since this is closed now, in case someone is still looking for a Vue compatible alternative you may want to try `knip`

**Pull request closed without merging**